revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,246 @@
+			return errToServerResult(server, nil)
+		}
+		return errToServerResult(server, RevokedError{})
+	default:
+		// ocsp.Unknown
+		return errToServerResult(server, UnknownStatusError{})
+	}
+}
+
+func executeOCSPCheck(cert, issuer *x509.Certificate, server string, opts Options) (*ocsp.Response, error) {
+	// TODO: Look into other alternatives for specifying the Hash
+	// https://github.com/notaryproject/notation-core-go/issues/139
+	ocspRequest, err := ocsp.CreateRequest(cert, issuer, &ocsp.RequestOptions{Hash: crypto.SHA256})
+	if err != nil {
+		return nil, OCSPCheckError{Err: err}
+	}
+
+	var resp *http.Response
+	if base64.URLEncoding.EncodedLen(len(ocspRequest)) >= 255 {
+		reader := bytes.NewReader(ocspRequest)
+		resp, err = opts.HTTPClient.Post(server, "application/ocsp-request", reader)
+	} else {
+		encodedReq := base64.URLEncoding.EncodeToString(ocspRequest)
+		var reqURL string
+		reqURL, err = url.JoinPath(server, encodedReq)
+		if err != nil {
+			return nil, OCSPCheckError{Err: err}
+		}
+		resp, err = opts.HTTPClient.Get(reqURL)
+	}
+
+	if err != nil {
+		var urlErr *url.Error
+		if errors.As(err, &urlErr) && urlErr.Timeout() {
+			return nil, TimeoutError{}
+		}
+		return nil, OCSPCheckError{Err: err}
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return nil, fmt.Errorf("failed to retrieve OSCP: response had status code %d", resp.StatusCode)
+	}
+
+	body, err := io.ReadAll(io.LimitReader(resp.Body, ocspMaxResponseSize))
+	if err != nil {
+		return nil, OCSPCheckError{Err: err}
+	}
+
+	switch {
+	case bytes.Equal(body, ocsp.UnauthorizedErrorResponse):
+		return nil, OCSPCheckError{Err: errors.New("OSCP unauthorized")}
+	case bytes.Equal(body, ocsp.MalformedRequestErrorResponse):
+		return nil, OCSPCheckError{Err: errors.New("OSCP malformed")}
+	case bytes.Equal(body, ocsp.InternalErrorErrorResponse):
+		return nil, OCSPCheckError{Err: errors.New("OSCP internal error")}
+	case bytes.Equal(body, ocsp.TryLaterErrorResponse):
+		return nil, OCSPCheckError{Err: errors.New("OSCP try later")}
+	case bytes.Equal(body, ocsp.SigRequredErrorResponse):
+		return nil, OCSPCheckError{Err: errors.New("OSCP signature required")}
+	}
+
+	return ocsp.ParseResponseForCert(body, cert, issuer)
+}
+
+func errToServerResult(server string, err error) *result.ServerResult {
+	if err == nil {
+		return &result.ServerResult{
+			Result: result.ResultOK,
+			Server: server,
+			Error:  nil,
+		}
+	} else if errors.Is(err, NoOCSPServerError{}) {
+		return &result.ServerResult{
+			Result: result.ResultNonRevokable,
+			Server: server,
+			Error:  nil,
+		}
+	} else if errors.Is(err, RevokedError{}) {
+		return &result.ServerResult{
+			Result: result.ResultRevoked,
+			Server: server,
+			Error:  err,
+		}
+	}
+	// Includes OCSPCheckError, UnknownStatusError, result.InvalidChainError, and
+	// TimeoutError
+	return &result.ServerResult{
+		Result: result.ResultUnknown,
+		Server: server,
+		Error:  err,
+	}
```

Comment:
nit this an be simplified by creating new method called NewServerResult in results.go

```
// String provides a conversion from a Result to a string
func (r ServerResult) New ServerResult(..., server string, err error) *ServerResult {
      return &result.ServerResult{
		Result: x,
		Server: y,
		Error:  z,
	}
}
```
```
	if err == nil {
		return NewServerResult{result.ResultOK, server, nil)
		}
	} else if errors.Is(err, NoOCSPServerError{}) {
		return NewServerResult{result.ResultNonRevokable, server, nil)
	} else if errors.Is(err, RevokedError{}) {
		return NewServerResult{result.ResultRevoked, server, err)
	}
	// Includes OCSPCheckError, UnknownStatusError, result.InvalidChainError, and
	// TimeoutError
        return NewServerResult{result.ResultUnknown, server, err)
```